concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+For the most part in computing, we tend to pretend and build abstractions as if
+numbers are [infinite scalars](#infinite-scalars).
+
+Of course, this is rarely true in practice.
+Integers are often implemented in such a way that values that would require beyond 32 (or 64, etc) bits are unrepresentable;
+arguably, this makes them more like a (very (*very*) large) enum than anything else.
+Floating point numbers implemented in IEE764 style also have their own entire host of foibles.
+
+The scope of this document is not to solve this conundrum, nor to try to convince you how you should consider numbers.
+However, in IPLD, we usually do consider integers to be [infinite scalars](#infinite-scalars),
+with the caveat that libraries processing data which exceeds the boundaries they can process should error and halt explicitly.
+(In IPLD, we have the good fortune that this definition is unusually easy to stick to:
+since we only carry data around, and don't provide mathematical operations on it,
+it's easy to expect that data which was processable on the way in will remain processable throughout the time we handle it.)
+
+
+Types in IPLD Schemas
+---------------------
+
+It probably won't surprise to you hear that IPLD Schema's core features
+map very closely to the categories of types we've discussed:
+
+- [enums](#enums) are `enum`.
+- [sum types](#sum-types) are `union`.
+- [product types](#product-types) are `struct`.
+- [infinite recursives](#infinite-recursives) appear as `list` and `map`.
+- [infinite scalars](#infinite-scalars) appear as `string`, `bytes`, and `int`.
+- [sum types](#sum-types) specifically for [maybe](#maybe-types) show up as `optional` and `nullable`.
+	- (there are two of them because each creates a sum with a different [unit type](#unit-types)!)
+
+IPLD Schemas don't have a concept of [interfaces](#open-vs-closed-membership).
+At least, *not exactly* -- not *within* the schema.
+Instead, the *whole Schema* is treated as an interface:
+since you can take some data and take a schema, and attempt to unify them to get typed data if the schema "matches"...
+you can use a series of Schemas in the same way as you'd use a series of interfaces.
+(IPLD Schemas are effectively [*structural*](explicit-versus-structural-interfaces) interfaces.)
+
+
+Different sizes of Infinity
+---------------------------
+
+// todo: not quite sure what to say about this.  the answer isn't "give up", that's for sure.
+
+// but i'm also not sure how to give this a brief treatment in a beginner-friendly way.
+
+// maybe just some admonishments about the dangers of subtracting from infinity -- i.e. "reserved keywords" -- would be useful.
+
+// "infinity plus one" is also still very useful to track opaquely as "a+1" rather than flatten to inf.
+
+// something about shifting/unshifting prefixes/suffixes could be useful -- link to "infinite hotel" metaphor.  folks seem to often find this a bit heady, though, so a more applied story would be nice.
```

Comment:
The infinite hotel, if I recall correctly, is about Cantor-style uncountables, not about enumerable cardinality. I would strongly advise against bringing this up in any way other than "if you've read the infinite hotel story, please don't try to apply that here".

Reply:
I think it still might apply to some situations?

Consider: I have two types, both of which are strings (so, they have cardinality `|inf|` and `|inf|`).  I want to make a new type which is a sum type of them.  If I still want to represent this as a string, I can actually do this losslessly... by prefixing them with a discriminator.  But it's essential to do that prefix consistently such that it's not subsumed in the infinity the two initial types can already contain or the result is insufficiently defined.


... you're right, it's still overkill, isn't it.  What I wanted to get at here is an admonition against trying to make sum types without discriminators, and especially doing it wrong in the form of sloppy manipulations of infinite scalars.  But that's a different thing and the infinite hotel certainly isn't the greatest tool for explaining this.

Anyway: I dropped this whole section, per other comments on it.  If it can't be finished today, it can be rebooted in some later PR.